### ISSUE TITLE
Ultimate does not support u'' and U'' type unicode chars and strings 

### ISSUE THREAD

**Issue by invd** [ultimate-pa/ultimate#322]
_Note: please add the appropriate labels such as feature/enhancement_.

Newer C standards are able to handle 16-bit and 32-bit unicode characters and strings. The following is valid with `gcc --pedantic`:
```
//#Safe
#include <uchar.h>

typedef short unsigned int char16_t;
typedef unsigned int char32_t;

int main(void) {
    char16_t c16 = u'\u00F6';
    char32_t c32 = U'\U0010FFFF';
}
```

Ultimate, however, is not able to model this kind of string yet:
```
 * Results from de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator:
  - ExceptionOrErrorResult: UnsupportedOperationException: null
    de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator: UnsupportedOperationException: null: de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.util.ISOIEC9899TC3.handleCharConstant(ISOIEC9899TC3.java:118)
```

This is relevant to the trezor verification as 16-bit strings are used in at least two instances, see [winusb_defs.h](https://github.com/trezor/trezor-mcu/blob/master/winusb_defs.h#L33) in the trezor-mcu repository.

The example code for this issue is copied from [LLVM Bug #24188](https://lists.llvm.org/pipermail/llvm-bugs/2015-July/040965.html), which is still unresolved and made me aware of this issue, as trezor-mcu is run with `-std=gnu99`, where clang fails to compile. For `-std=gnu11`, clang appears to be handling such unicode literals similarly well as gcc in it's default settings, so in my opinion, it is valuable to include this support in Ultimate although not all compilers accept it for all C standards.

Further information:
* Ultimate 0.1.23-6f35ba1
* AutomizerMemDerefMemtrack.xml 
* svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf

**Comment by invd** [comment]
The current Ultimate 0.1.23-3c47570 no longer throws the `UnsupportedOperationException: null` when verifying the example program mentioned above. I won't have the time to go into further tests for correct access / output behavior, but this looks very promising. Thanks for the quick solution!

**Comment by Heizmann** [comment]
I did not yet implement proper support for unicode strings and I would prefer to postpone that because it would require a further refactoring of the way how the underlying Boogie program writes strings to the heap array.

**Comment by invd** [comment]
At the moment, the program subsets given to Ultimate do not include the winusb_defs.h-section, so I can't say with certainty that the current level of support will be enough for the verification coverage we're hoping for. I will update this ticket once more information on this is available.

**Comment by invd** [comment]
Due to the (unrelated) decision to revert certain WebUSB-changes in the trezor-mcu repository, the problematic winfiles are excluded from the project for the moment, see [the relevant commit](https://github.com/trezor/trezor-mcu/commit/b20336e82d19d95a0ce65d13d6c7bcf9d11ac668). Consequently, this change lowers the priority of the current issue, as we're likely not blocked by soundness issues relating to this in the upcoming weeks, for example.

**Comment by invd** [comment]
Today's Ultimate runs over the example program listed above, as expected:

`./Ultimate -tc ~/ultimate/releaseScripts/default/UAutomizer-linux/config/AutomizerMemDerefMemtrack.xml -s ~/ultimate/trunk/examples/settings/default/automizer/svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf -i ~/ultimate/trunk/examples/CToBoogieTranslation/trezor/unicode_u16_u32_string_test.c`
`RESULT: Ultimate proved your program to be correct!`

However, as I understand the current situation, Ultimate still has not come closer to really supporting Unicode strings, so this issue should stay open.

**Comment by invd** [comment]
@Heizmann: what's the status on this?
Ultimate can translate the [unicode_u16_u32_string_test.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/unicode_u16_u32_string_test.c) test file for a while now, but is there noteworthy progress beyond that?

**Comment by Heizmann** [comment]
@invd Unfortunately, there is no progress here. I could implement support for unicode strings but then I would have to postpone other features.
